### PR TITLE
DrivAerML: no-EMA + lr=8e-4 + cosine T_max sweep (10/20/30)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Shorter cosine T_max may help DrivAerML as it helped TandemFoil.** T_max=30 was optimal for TandemFoil (#2435). The DrivAerML baseline (#2467) uses T_max=150, which barely moves the LR in 2 epochs. Shorter T_max values complete LR cycles within the budget and may find better minima. Combined with the proven no-EMA + lr=8e-4 recipe.

## Current Baseline

| Dataset | Metric | Value | PR |
|---|---|---|---|
| DrivAerML | val_primary/surface_rel_l2_pct | **56.91%** | #2467 (no-EMA + AdamW lr=8e-4, T_max=150) |

## Design

Three runs varying T_max with the winning config:

| Run | T_max | wandb_name |
|---|---|---|
| 1 | 10 | nezuko-noema-tmax10 |
| 2 | 20 | nezuko-noema-tmax20 |
| 3 | 30 | nezuko-noema-tmax30 |

## Instructions to Student

Run **sequentially** (one at a time):

```bash
cd target/icml2026

# Run 1: T_max=10
python train.py --dataset drivaerml --optimizer adamw --lr 8e-4 --cosine_t_max 10 \
  --no-use-ema \
  --wandb_group nezuko-drivaerml-tmax --wandb_name nezuko-noema-tmax10

# Run 2: T_max=20
python train.py --dataset drivaerml --optimizer adamw --lr 8e-4 --cosine_t_max 20 \
  --no-use-ema \
  --wandb_group nezuko-drivaerml-tmax --wandb_name nezuko-noema-tmax20

# Run 3: T_max=30
python train.py --dataset drivaerml --optimizer adamw --lr 8e-4 --cosine_t_max 30 \
  --no-use-ema \
  --wandb_group nezuko-drivaerml-tmax --wandb_name nezuko-noema-tmax30
```

**Report** `val_primary/surface_rel_l2_pct` and epochs per run. Target: beat 56.91%.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Run 1: T_max=10

| Metric | Epoch 1 | Epoch 2 |
|---|---|---|
| val_primary/surface_rel_l2_pct | 64.33% | **56.51%** |
| loss | 0.870 | 0.670 |
| lr | 2.76e-4 | 7.64e-5 |
| test_primary/surface_rel_l2_pct | — | **58.33%** |

- **W&B run:** `5qap4o8v` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/5qap4o8v))
- **Epochs completed:** 2/2

### Run 2: T_max=20

| Metric | Epoch 1 | Epoch 2 |
|---|---|---|
| val_primary/surface_rel_l2_pct | 66.87% | **59.13%** |
| loss | 0.905 | 0.694 |
| lr | 6.35e-4 | 2.76e-4 |
| test_primary/surface_rel_l2_pct | — | **58.96%** |

- **W&B run:** `2q6vkrsy` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/2q6vkrsy))
- **Epochs completed:** 2/2

### Run 3: T_max=30

| Metric | Epoch 1 | Epoch 2 |
|---|---|---|
| val_primary/surface_rel_l2_pct | 64.78% | **58.72%** |
| loss | 0.874 | 0.695 |
| lr | 3.46e-5 | 6.68e-4 |
| test_primary/surface_rel_l2_pct | — | **60.17%** |

- **W&B run:** `4f9n1032` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/4f9n1032))
- **Epochs completed:** 2/2

### Comparison against baseline

| Run | T_max | val_primary | test_primary | Delta val vs baseline |
|---|---|---|---|---|
| **Baseline (#2467)** | 150 | 56.91% | 57.33% | — |
| **T_max=10** | 10 | **56.51%** | 58.33% | **-0.40 pp** |
| **T_max=20** | 20 | 59.13% | 58.96% | +2.22 pp |
| **T_max=30** | 30 | 58.72% | 60.17% | +1.81 pp |

### Command used

```bash
# All runs used the same base config, varying only --cosine-t-max:
CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 8e-4 \
  --cosine-t-max {10,20,30} --no-use-ema --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent nezuko --wandb-name "nezuko-noema-tmax{N}" --wandb-group "nezuko-drivaerml-tmax"
```

- **Peak memory:** Fits within a single 95 GiB GPU. Note: `--drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --batch-size 1` required to avoid OOM (same as baseline #2467).

### What happened

**T_max=10 is the only variant that beats baseline on val** (56.51% vs 56.91%, -0.40 pp), but it's slightly worse on test (58.33% vs 57.33%, +1.00 pp). T_max=20 and T_max=30 are clearly worse on val.

The key insight is that **shorter T_max values with step-level scheduling create very different LR dynamics at 2 epochs**:
- **T_max=10:** With 394 steps/epoch, this completes ~39 full cosine cycles per epoch. The rapid cycling creates a warm-restart-like effect, keeping the optimizer exploring. This seems to help val convergence.
- **T_max=20:** Completes ~20 cycles/epoch. The intermediate cycling doesn't help — epoch 1 val (66.87%) is the worst of the three.
- **T_max=30:** Completes ~13 cycles/epoch. Similar to T_max=20, no clear benefit.
- **T_max=150 (baseline):** Completes ~2.6 cycles/epoch. The slower decay keeps a higher effective LR longer, which appears to help test generalization.

The val improvement from T_max=10 is marginal (0.40 pp) and doesn't transfer to test, suggesting the rapid cycling may slightly overfit to val. The baseline T_max=150 remains the strongest overall config when considering both val and test.

### Suggested follow-ups

1. **More epochs** — all runs are still converging at epoch 2. With 4-6 epochs, the T_max differences might separate more clearly since shorter cycles would complete more full restarts.
2. **T_max=5** — if very short cycles help, pushing even shorter might amplify the effect.
3. **Epoch-level T_max** — the current T_max is step-level. An epoch-level cosine (T_max_epochs) might be more meaningful for DrivAerML where each epoch is ~394 steps.
4. **Focus on other axes** — since T_max tuning yields only marginal gains on DrivAerML, other approaches (Fourier features, model capacity, more training time) may be higher leverage.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2480 because ADVISOR: Closing — T_max=10 val=56.51% barely beats 56.91% but test=58.33% is worse (regression). T_max=20 and T_max=30 both clearly worse. The val improvement doesn't generalize — likely val overfitting to cosine cycling pattern. Baseline T_max=150 remains best on the test metric. NEW: Fourier features just pushed DrivAerML to 54.33% (PR #2479). Redirecting you to build on that result..
